Support for compiling own (test) code with Groovy 4.0

### Expected Behavior
Gradle should allow to use Groovy 4 to **compile** Groovy code (just compile my code, NOT to use Groovy 4 to build Gradle).

### Current Behavior
Due to changing the groupId in Groovy 4 from `org.codehaus.groovy` to `org.apache.groovy`, Gradle is not able to properly find and use Groovy 4 JARs.

### Context
As a Spock developer, I would like to be able to check if Spock 2 works properly with Groovy 4 (even with just 4.0.0-SNAPSHOT) to provide feedback for the Groovy team. Unfortunately the compilation is not possible due to limitations in Gradle:

```
> Task :compileGroovy FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':compileGroovy'.
> Could not resolve all files for configuration ':detachedConfiguration1'.
   > Could not find org.codehaus.groovy:groovy:4.0.0-SNAPSHOT.
     Searched in the following locations:
       - https://repo.maven.apache.org/maven2/org/codehaus/groovy/groovy/4.0.0-SNAPSHOT/maven-metadata.xml
       - https://repo.maven.apache.org/maven2/org/codehaus/groovy/groovy/4.0.0-SNAPSHOT/groovy-4.0.0-SNAPSHOT.pom
       - https://oss.jfrog.org/artifactory/oss-snapshot-local/org/codehaus/groovy/groovy/4.0.0-SNAPSHOT/maven-metadata.xml
       - https://oss.jfrog.org/artifactory/oss-snapshot-local/org/codehaus/groovy/groovy/4.0.0-SNAPSHOT/groovy-4.0.0-SNAPSHOT.pom
     Required by:
         project :
   > Could not find org.codehaus.groovy:groovy-ant:4.0.0-SNAPSHOT.
     Searched in the following locations:
       - https://repo.maven.apache.org/maven2/org/codehaus/groovy/groovy-ant/4.0.0-SNAPSHOT/maven-metadata.xml
       - https://repo.maven.apache.org/maven2/org/codehaus/groovy/groovy-ant/4.0.0-SNAPSHOT/groovy-ant-4.0.0-SNAPSHOT.pom
       - https://oss.jfrog.org/artifactory/oss-snapshot-local/org/codehaus/groovy/groovy-ant/4.0.0-SNAPSHOT/maven-metadata.xml
       - https://oss.jfrog.org/artifactory/oss-snapshot-local/org/codehaus/groovy/groovy-ant/4.0.0-SNAPSHOT/groovy-ant-4.0.0-SNAPSHOT.pom
     Required by:
         project :
   > Could not find org.codehaus.groovy:groovy-templates:4.0.0-SNAPSHOT.
     Searched in the following locations:
       - https://repo.maven.apache.org/maven2/org/codehaus/groovy/groovy-templates/4.0.0-SNAPSHOT/maven-metadata.xml
       - https://repo.maven.apache.org/maven2/org/codehaus/groovy/groovy-templates/4.0.0-SNAPSHOT/groovy-templates-4.0.0-SNAPSHOT.pom
       - https://oss.jfrog.org/artifactory/oss-snapshot-local/org/codehaus/groovy/groovy-templates/4.0.0-SNAPSHOT/maven-metadata.xml
       - https://oss.jfrog.org/artifactory/oss-snapshot-local/org/codehaus/groovy/groovy-templates/4.0.0-SNAPSHOT/groovy-templates-4.0.0-SNAPSHOT.pom
     Required by:
         project :
```

It seems that Gradle has `org.codehaus.groovy` hardcoded that this causes that even there is [`org.apache.groovy:groovy:4.0.0-SNAPSHOT`](https://oss.jfrog.org/artifactory/oss-snapshot-local/org/apache/groovy/) as a dependency, it tries to find `org.codehaus.groovy:groovy:4.0.0-SNAPSHOT` (also groovy-ant and groovy-template) with wrong groupId which doesn't exist.

Sample configuration from generated Gradle project (in Kotlin to do not complicate with Groovy as the DSL language :) ):

```
plugins {
    groovy
    `java-library`
}

repositories {
    mavenCentral()
    maven { url = uri("https://oss.jfrog.org/artifactory/oss-snapshot-local/") }
}

dependencies {
//    implementation("org.codehaus.groovy:groovy:3.0.4")
    implementation("org.apache.groovy:groovy:4.0.0-SNAPSHOT")

    testImplementation("org.spockframework:spock-core:2.0-M3-groovy-3.0") {
        exclude(group = "org.codehaus.groovy")  //to eliminate "old" 3.x groovy.jar
    }
}

tasks {
    withType<Test>().configureEach {
        useJUnitPlatform()
    }
}
```
(it is expected to have this failing due to Spock and Groovy version mismatch, but it is possible to disabled the check with [`-Dspock.iKnowWhatImDoing.disableGroovyVersionCheck=true`](http://spockframework.org/spock/docs/2.0-M3/known_issues.html#_groovy_version_compatibility), so this has a chance to work - once Gradle will be able to find Groovy 4 with its changed groupId).

GitHub: https://github.com/szpak/code-examples-and-poc/blob/master/groovy4-gradle-kotlin/build.gradle.kts

As I was talking with Paul King recently, there are **no plans** to change also package for Groovy 4 (which would be even more problematic).

Tested with Gradle 6.5.

--- a/subprojects/plugins/src/main/java/org/gradle/api/internal/plugins/GroovyJarFile.java
+++ b/subprojects/plugins/src/main/java/org/gradle/api/internal/plugins/GroovyJarFile.java
@@ -54,11 +54,15 @@ public class GroovyJarFile {
     }
 
     public String getDependencyNotation() {
-        String result = "org.codehaus.groovy:" + getBaseName() + ":" + getVersion();
+        String result = getGroupId() + ":" + getBaseName() + ":" + getVersion();
         if (isIndy()) {
             result += ":indy";
         }
         return result;
+    }
+
+    private String getGroupId() {
+        return getVersion().getMajor() < 4 ? "org.codehaus.groovy" : "org.apache.groovy";
     }
 
     @Nullable

--- a/subprojects/plugins/src/test/groovy/org/gradle/api/internal/plugins/GroovyJarFileTest.groovy
+++ b/subprojects/plugins/src/test/groovy/org/gradle/api/internal/plugins/GroovyJarFileTest.groovy
@@ -64,4 +64,71 @@ class GroovyJarFileTest extends Specification {
         jar.indy
         jar.dependencyNotation == "org.codehaus.groovy:groovy:2.0.5:indy"
     }
+
+    def "parse 'groovy' Jar for version 4 has apache groupId"() {
+        def jar = GroovyJarFile.parse(new File("/lib/groovy-4.0.0.jar"))
+
+        expect:
+        jar != null
+        jar.file == new File("/lib/groovy-4.0.0.jar")
+        jar.baseName == "groovy"
+        jar.version == VersionNumber.parse("4.0.0")
+        !jar.groovyAll
+        !jar.indy
+        jar.dependencyNotation == "org.apache.groovy:groovy:4.0.0"
+    }
+
+    def "parse 'groovy' Jar for version above 4 has apache groupId"() {
+        def jar = GroovyJarFile.parse(new File("/lib/groovy-5.0.1.jar"))
+
+        expect:
+        jar != null
+        jar.file == new File("/lib/groovy-5.0.1.jar")
+        jar.baseName == "groovy"
+        jar.version == VersionNumber.parse("5.0.1")
+        !jar.groovyAll
+        !jar.indy
+        jar.dependencyNotation == "org.apache.groovy:groovy:5.0.1"
+    }
+
+    def "parse 'groovy-all' Jar for version 4 has apache groupId"() {
+        def jar = GroovyJarFile.parse(new File("/lib/groovy-all-4.0.5.jar"))
+
+        expect:
+        jar != null
+        jar.file == new File("/lib/groovy-all-4.0.5.jar")
+        jar.baseName == "groovy-all"
+        jar.version == VersionNumber.parse("4.0.5")
+        jar.groovyAll
+        !jar.indy
+        jar.dependencyNotation == "org.apache.groovy:groovy-all:4.0.5"
+
+    }
+
+    def "parse 'groovy-all' Jar for version above 4 has apache groupId"() {
+        def jar = GroovyJarFile.parse(new File("/lib/groovy-all-5.0.5.jar"))
+
+        expect:
+        jar != null
+        jar.file == new File("/lib/groovy-all-5.0.5.jar")
+        jar.baseName == "groovy-all"
+        jar.version == VersionNumber.parse("5.0.5")
+        jar.groovyAll
+        !jar.indy
+        jar.dependencyNotation == "org.apache.groovy:groovy-all:5.0.5"
+
+    }
+
+    def "parse indy Jar for version 4 has apache groupId"() {
+        def jar = GroovyJarFile.parse(new File("/lib/groovy-4.0.5-indy.jar"))
+
+        expect:
+        jar != null
+        jar.file == new File("/lib/groovy-4.0.5-indy.jar")
+        jar.baseName == "groovy"
+        jar.version == VersionNumber.parse("4.0.5")
+        !jar.groovyAll
+        jar.indy
+        jar.dependencyNotation == "org.apache.groovy:groovy:4.0.5:indy"
+    }
 }

--- a/subprojects/plugins/src/test/groovy/org/gradle/api/internal/plugins/GroovyJarFileTest.groovy
+++ b/subprojects/plugins/src/test/groovy/org/gradle/api/internal/plugins/GroovyJarFileTest.groovy
@@ -25,110 +25,65 @@ class GroovyJarFileTest extends Specification {
         GroovyJarFile.parse(new File("groovy-2.0.5.zip")) == null
     }
 
-    def "parse 'groovy' Jar"() {
-        def jar = GroovyJarFile.parse(new File("/lib/groovy-2.0.5.jar"))
+    def "parse 'groovy' Jar"(String fileName, String version, String dependencyNotation) {
+        def jar = GroovyJarFile.parse(new File("$fileName"))
 
         expect:
         jar != null
-        jar.file == new File("/lib/groovy-2.0.5.jar")
+        jar.file == new File("$fileName")
         jar.baseName == "groovy"
-        jar.version == VersionNumber.parse("2.0.5")
+        jar.version == VersionNumber.parse("$version")
         !jar.groovyAll
         !jar.indy
-        jar.dependencyNotation == "org.codehaus.groovy:groovy:2.0.5"
+        jar.dependencyNotation == "$dependencyNotation"
+
+        where:
+        fileName                | version | dependencyNotation
+        "/lib/groovy-2.0.5.jar" | "2.0.5" | "org.codehaus.groovy:groovy:2.0.5"
+        "/lib/groovy-4.0.0.jar" | "4.0.0" | "org.apache.groovy:groovy:4.0.0"
+        "/lib/groovy-4.1.0.jar" | "4.1.0" | "org.apache.groovy:groovy:4.1.0"
+        "/lib/groovy-5.0.2.jar" | "5.0.2" | "org.apache.groovy:groovy:5.0.2"
     }
 
-    def "parse 'groovy-all' Jar"() {
-        def jar = GroovyJarFile.parse(new File("/lib/groovy-all-2.0.5.jar"))
+    def "parse 'groovy-all' Jar"(String fileName, String version, String dependencyNotation) {
+        def jar = GroovyJarFile.parse(new File("$fileName"))
 
         expect:
         jar != null
-        jar.file == new File("/lib/groovy-all-2.0.5.jar")
+        jar.file == new File("$fileName")
         jar.baseName == "groovy-all"
-        jar.version == VersionNumber.parse("2.0.5")
+        jar.version == VersionNumber.parse("$version")
         jar.groovyAll
         !jar.indy
-        jar.dependencyNotation == "org.codehaus.groovy:groovy-all:2.0.5"
+        jar.dependencyNotation == "$dependencyNotation"
 
+        where:
+        fileName                    | version | dependencyNotation
+        "/lib/groovy-all-2.0.5.jar" | "2.0.5" | "org.codehaus.groovy:groovy-all:2.0.5"
+        "/lib/groovy-all-4.0.0.jar" | "4.0.0" | "org.apache.groovy:groovy-all:4.0.0"
+        "/lib/groovy-all-4.1.0.jar" | "4.1.0" | "org.apache.groovy:groovy-all:4.1.0"
+        "/lib/groovy-all-5.0.2.jar" | "5.0.2" | "org.apache.groovy:groovy-all:5.0.2"
     }
 
-    def "parse indy Jar"() {
-        def jar = GroovyJarFile.parse(new File("/lib/groovy-2.0.5-indy.jar"))
+    def "parse indy Jar"(String fileName, String version, String dependencyNotation) {
+        def jar = GroovyJarFile.parse(new File("$fileName"))
 
         expect:
         jar != null
-        jar.file == new File("/lib/groovy-2.0.5-indy.jar")
+        jar.file == new File("$fileName")
         jar.baseName == "groovy"
-        jar.version == VersionNumber.parse("2.0.5")
+        jar.version == VersionNumber.parse("$version")
         !jar.groovyAll
         jar.indy
-        jar.dependencyNotation == "org.codehaus.groovy:groovy:2.0.5:indy"
+        jar.dependencyNotation == "$dependencyNotation"
+
+        where:
+        fileName                     | version | dependencyNotation
+        "/lib/groovy-2.0.5-indy.jar" | "2.0.5" | "org.codehaus.groovy:groovy:2.0.5:indy"
+        "/lib/groovy-4.0.0-indy.jar" | "4.0.0" | "org.apache.groovy:groovy:4.0.0:indy"
+        "/lib/groovy-4.1.0-indy.jar" | "4.1.0" | "org.apache.groovy:groovy:4.1.0:indy"
+        "/lib/groovy-5.0.2-indy.jar" | "5.0.2" | "org.apache.groovy:groovy:5.0.2:indy"
     }
 
-    def "parse 'groovy' Jar for version 4 has apache groupId"() {
-        def jar = GroovyJarFile.parse(new File("/lib/groovy-4.0.0.jar"))
 
-        expect:
-        jar != null
-        jar.file == new File("/lib/groovy-4.0.0.jar")
-        jar.baseName == "groovy"
-        jar.version == VersionNumber.parse("4.0.0")
-        !jar.groovyAll
-        !jar.indy
-        jar.dependencyNotation == "org.apache.groovy:groovy:4.0.0"
-    }
-
-    def "parse 'groovy' Jar for version above 4 has apache groupId"() {
-        def jar = GroovyJarFile.parse(new File("/lib/groovy-5.0.1.jar"))
-
-        expect:
-        jar != null
-        jar.file == new File("/lib/groovy-5.0.1.jar")
-        jar.baseName == "groovy"
-        jar.version == VersionNumber.parse("5.0.1")
-        !jar.groovyAll
-        !jar.indy
-        jar.dependencyNotation == "org.apache.groovy:groovy:5.0.1"
-    }
-
-    def "parse 'groovy-all' Jar for version 4 has apache groupId"() {
-        def jar = GroovyJarFile.parse(new File("/lib/groovy-all-4.0.5.jar"))
-
-        expect:
-        jar != null
-        jar.file == new File("/lib/groovy-all-4.0.5.jar")
-        jar.baseName == "groovy-all"
-        jar.version == VersionNumber.parse("4.0.5")
-        jar.groovyAll
-        !jar.indy
-        jar.dependencyNotation == "org.apache.groovy:groovy-all:4.0.5"
-
-    }
-
-    def "parse 'groovy-all' Jar for version above 4 has apache groupId"() {
-        def jar = GroovyJarFile.parse(new File("/lib/groovy-all-5.0.5.jar"))
-
-        expect:
-        jar != null
-        jar.file == new File("/lib/groovy-all-5.0.5.jar")
-        jar.baseName == "groovy-all"
-        jar.version == VersionNumber.parse("5.0.5")
-        jar.groovyAll
-        !jar.indy
-        jar.dependencyNotation == "org.apache.groovy:groovy-all:5.0.5"
-
-    }
-
-    def "parse indy Jar for version 4 has apache groupId"() {
-        def jar = GroovyJarFile.parse(new File("/lib/groovy-4.0.5-indy.jar"))
-
-        expect:
-        jar != null
-        jar.file == new File("/lib/groovy-4.0.5-indy.jar")
-        jar.baseName == "groovy"
-        jar.version == VersionNumber.parse("4.0.5")
-        !jar.groovyAll
-        jar.indy
-        jar.dependencyNotation == "org.apache.groovy:groovy:4.0.5:indy"
-    }
 }

